### ISSUE TITLE
Update items calling SaveChanges multiple times causes problems

### ISSUE THREAD

**Issue by lucamorelli** [dotnet/efcore#8821]
I need to access a database using EF Core, reading data from a `DbContext` and writing to another, using **two different instances of the same `DbContext`**, that are shared among services for all the request call.

Since the application it's quite complex and inside the same request the processing spans across multiple services, happens that sometime a procedure read an entity from the database, update it and save changes.

Considering that the consistency of the read and updates it's not a problem, I have a problem with the change tracker:

The first time a read an entity, update and save everything it's ok, but if I need to read again the same entity a second time, update and save, I have a problem.

I've checked and the problem is that after the first call to `SaveChanges`, the change tracker resets the enities to `unchanged` state, don't reset the content of the change tracker so, when I add to the context a second entity with the same key, EF tells me that this entity is already tracked.

Actually I added this check and seems to work:

```
foreach (var entry in db.ChangeTracker.Entries<UTENTI_FEA>())
{
  if ((int)entry.Property("UID").CurrentValue == fea.UID)
  {
     entry.State = EntityState.Detached;
     break;
  }
}
```
This way, if present, I detach the old entities before to add the new one.
Is there a better way to avoid the problem?

On StackOverflow has been suggested to use [DbContextScope](https://github.com/mehdime/DbContextScope), but I would like to know if there is some other solutions and/or this this scenario could be handled by EF.

**Comment by ajcvickers** [comment]
@lucamorelli It's hard to follow exactly what you are doing, but it is usually best to only keep a context instance around for a single unit-of-work. So it seems likely that in this case the context used for saving should be thrown away after SaveChanges has been called. Also, you might want to look at using AsNoTracking for your queries on the first context if you're not using this context to actually track changes.

**Comment by lucamorelli** [comment]
I'm trying to use a CQRS like approach, separating the read/query part from the write/command part of the data access, so I can change the source and or destination of data if needed indipendently.
I would like to see if I can share a single dbcontext along all the request as usual, because most of the time it works fine, but in a few cases the logic becomes complex, and can't be sure that just one SaveChanges is called.
Anyway looks like doesn't work:I'll look if fix as shown above or try this DbContextScope, because adapt the code to use a single unit of work could be source of compatibility problems
